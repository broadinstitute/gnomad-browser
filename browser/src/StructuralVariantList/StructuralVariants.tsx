import React, { useCallback, useMemo, useRef, useState } from 'react'
import styled from 'styled-components'

// @ts-expect-error TS(7016) FIXME: Could not find a declaration file for module '@gno... Remove this comment to see the full error message
import { PositionAxisTrack } from '@gnomad/region-viewer'
import { Button, SegmentedControl } from '@gnomad/ui'

import { showNotification } from '../Notifications'
import StatusMessage from '../StatusMessage'
import { TrackPageSection } from '../TrackPage'
import userPreferences from '../userPreferences'
import VariantTableConfigurationModal from '../VariantList/VariantTableConfigurationModal'

import ExportStructuralVariantsButton from './ExportStructuralVariantsButton'
import filterStructuralVariants from './filterStructuralVariants'
import {
  svConsequenceCategories,
  svConsequenceCategoryColors,
} from './structuralVariantConsequences'
import { svTypeColors } from './structuralVariantTypes'
import StructuralVariantFilterControls from './StructuralVariantFilterControls'
import StructrualVariantPropType from './StructuralVariantPropType'
import structuralVariantTableColumns, {
  getColumnsForContext,
} from './structuralVariantTableColumns'
import StructuralVariantsTable from './StructuralVariantsTable'
import StructuralVariantTracks from './StructuralVariantTracks'

const NUM_ROWS_RENDERED = 20
const TRACK_HEIGHT = 14
const TABLE_ROW_HEIGHT = 25

const Wrapper = styled.div`
  margin-bottom: 1em;
`

const ControlWrapper = styled(Wrapper)`
  display: flex;
  justify-content: flex-end;
  align-items: center;

  @media (min-width: 900px) {
    margin-right: 160px;
  }
`

const HUMAN_CHROMOSOMES = [...Array.from(new Array(22), (x: any, i: any) => `${i + 1}`), 'X', 'Y']

const DEFAULT_COLUMNS = [
  'source',
  'consequence',
  'class',
  'pos',
  'length',
  'ac',
  'an',
  'af',
  'homozygote_count',
]

const sortVariants = (variants: any, { sortKey, sortOrder }: any) => {
  const sortColumn = structuralVariantTableColumns.find((column: any) => column.key === sortKey)
  // @ts-expect-error TS(2532) FIXME: Object is possibly 'undefined'.
  return [...variants].sort((v1, v2) => sortColumn.compareFunction(v1, v2, sortOrder))
}

export interface Context {
  chrom: string
}

type StructuralVariantsProps = {
  context: Context
  exportFileName: string
  variants: StructrualVariantPropType[]
}

const StructuralVariants = ({ context, exportFileName, variants }: StructuralVariantsProps) => {
  const table = useRef(null)
  const tracks = useRef(null)

  const [selectedColumns, setSelectedColumns] = useState(() => {
    try {
      return userPreferences.getPreference('structuralVariantTableColumns') || DEFAULT_COLUMNS
    } catch (error) {
      return DEFAULT_COLUMNS
    }
  })

  const renderedTableColumns = useMemo(() => {
    const columnsForContext = getColumnsForContext(context)
    return (
      ['variant_id', ...selectedColumns]
        // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        .map((columnKey) => columnsForContext[columnKey])
        .filter(Boolean)
        .map((column) => ({
          ...column,
          isSortable: Boolean(column.compareFunction),
          tooltip: column.description,
        }))
    )
  }, [context, selectedColumns])

  const [filter, setFilter] = useState({
    includeConsequenceCategories: {
      lof: true,
      dup_lof: true,
      copy_gain: true,
      other: true,
    },
    includeTypes: {
      DEL: true,
      DUP: true,
      MCNV: true,
      INS: true,
      INV: true,
      CPX: true,
      OTH: true,
    },
    includeFilteredVariants: false,
    searchText: '',
  })

  const [sortState, setSortState] = useState({
    sortKey: 'variant_id',
    sortOrder: 'ascending',
  })
  const { sortKey, sortOrder } = sortState

  // @ts-expect-error TS(7006) FIXME: Parameter 'newSortKey' implicitly has an 'any' typ... Remove this comment to see the full error message
  const setSortKey = useCallback((newSortKey) => {
    setSortState((prevSortState) => {
      if (newSortKey === prevSortState.sortKey) {
        return {
          sortKey: newSortKey,
          sortOrder: prevSortState.sortOrder === 'ascending' ? 'descending' : 'ascending',
        }
      }

      return {
        sortKey: newSortKey,
        sortOrder: 'descending',
      }
    })
  }, [])

  const filteredVariants = useMemo(
    () => filterStructuralVariants(variants, filter, renderedTableColumns),
    [variants, filter, renderedTableColumns]
  )
  const renderedVariants = useMemo(
    () => sortVariants(filteredVariants, sortState),
    [filteredVariants, sortState]
  )

  const [showTableConfigurationModal, setShowTableConfigurationModal] = useState(false)
  const [variantHoveredInTable, setVariantHoveredInTable] = useState(null)
  const [variantHoveredInTrack, setVariantHoveredInTrack] = useState(null)

  const shouldHighlightTableRow = useCallback(
    // @ts-expect-error TS(7006) FIXME: Parameter 'variant' implicitly has an 'any' type.
    (variant) => {
      return variant.variant_id === variantHoveredInTrack
    },
    [variantHoveredInTrack]
  )

  // @ts-expect-error TS(7031) FIXME: Binding element 'scrollOffset' implicitly has an '... Remove this comment to see the full error message
  const onScrollTable = useCallback(({ scrollOffset, scrollUpdateWasRequested }) => {
    if (tracks.current && !scrollUpdateWasRequested) {
      ;(tracks.current as any).scrollTo(
        Math.round(scrollOffset * (TRACK_HEIGHT / TABLE_ROW_HEIGHT))
      )
    }
  }, [])

  // @ts-expect-error TS(7031) FIXME: Binding element 'scrollOffset' implicitly has an '... Remove this comment to see the full error message
  const onScrollTracks = useCallback(({ scrollOffset, scrollUpdateWasRequested }) => {
    if (table.current && !scrollUpdateWasRequested) {
      ;(table.current as any).scrollTo(Math.round(scrollOffset * (TABLE_ROW_HEIGHT / TRACK_HEIGHT)))
    }
  }, [])

  const [colorKey, setColorKey] = useState('type')
  const trackColor = useCallback(
    // @ts-expect-error TS(7006) FIXME: Parameter 'variant' implicitly has an 'any' type.
    (variant) => {
      if (colorKey === 'type') {
        // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
        return svTypeColors[variant.type] || svTypeColors.OTH
      }
      return variant.consequence
        ? // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
          svConsequenceCategoryColors[svConsequenceCategories[variant.consequence]]
        : svConsequenceCategoryColors.other
    },
    [colorKey]
  )

  if (variants.length === 0) {
    return <StatusMessage>No variants found</StatusMessage>
  }

  const numRowsRendered = Math.min(renderedVariants.length, NUM_ROWS_RENDERED)

  // pos/end and pos2/end2 coordinates are based on the chromosome which they are located on.
  // If that chromosome is not the same as the one that the region viewer's coordinates
  // are based on, then offset the positions so that they are based on the
  // region viewer's coordinate system.
  const currentChromIndex = HUMAN_CHROMOSOMES.indexOf(context.chrom) // eslint-disable-line react/destructuring-assignment
  const positionCorrectedVariants = renderedVariants.map((variant) => {
    const copy = { ...variant }

    // This can only happen when chrom2/pos2/end2 is non-null
    if (variant.chrom2) {
      const chromIndex = HUMAN_CHROMOSOMES.indexOf(variant.chrom)
      const endChromIndex = HUMAN_CHROMOSOMES.indexOf(variant.chrom2)

      copy.pos += (chromIndex - currentChromIndex) * 1e9
      copy.end += (chromIndex - currentChromIndex) * 1e9

      copy.pos2 += (endChromIndex - currentChromIndex) * 1e9
      copy.end2 += (endChromIndex - currentChromIndex) * 1e9
    }

    return copy
  })

  return (
    <div>
      <ControlWrapper>
        <span style={{ marginRight: '0.5em' }}>Color variants by</span>
        <SegmentedControl
          id="sv-color-key"
          options={[
            { label: 'Consequence', value: 'consequence' },
            { label: 'Class', value: 'type' },
          ]}
          value={colorKey}
          onChange={setColorKey}
        />
      </ControlWrapper>
      <Wrapper>
        <StructuralVariantTracks
          ref={tracks}
          // @ts-expect-error TS(2322) FIXME: Type '{ ref: MutableRefObject<null>; highlightedVa... Remove this comment to see the full error message
          highlightedVariant={variantHoveredInTable}
          numTracksRendered={numRowsRendered}
          onHover={setVariantHoveredInTrack}
          onScroll={onScrollTracks}
          trackColor={trackColor}
          trackHeight={TRACK_HEIGHT}
          variants={positionCorrectedVariants}
        />
      </Wrapper>
      <Wrapper>
        <PositionAxisTrack />
      </Wrapper>
      <TrackPageSection style={{ fontSize: '14px' }}>
        <Wrapper>
          <StructuralVariantFilterControls
            // @ts-expect-error TS(2322) FIXME: Type 'string' is not assignable to type '"type" | ... Remove this comment to see the full error message
            colorKey={colorKey}
            value={filter}
            onChange={setFilter}
          />
          <div>
            <ExportStructuralVariantsButton
              exportFileName={exportFileName}
              variants={renderedVariants}
            />

            <Button
              onClick={() => {
                setShowTableConfigurationModal(true)
              }}
              style={{ marginLeft: '1ch' }}
            >
              Configure table
            </Button>
          </div>
        </Wrapper>
        <Wrapper
          style={{
            // Keep the height of the table section constant when filtering variants, avoid layout shift
            minHeight: 40 + TABLE_ROW_HEIGHT * Math.min(variants.length, NUM_ROWS_RENDERED),
          }}
        >
          {renderedVariants.length ? (
            <StructuralVariantsTable
              ref={table}
              // @ts-expect-error TS(2322) FIXME: Type '{ ref: MutableRefObject<null>; cellData: { c... Remove this comment to see the full error message
              cellData={{
                colorKey,
                highlightWords: filter.searchText.split(',').map((s) => s.trim()),
              }}
              columns={renderedTableColumns}
              numRowsRendered={numRowsRendered}
              onHoverVariant={setVariantHoveredInTable}
              onRequestSort={setSortKey}
              onScroll={onScrollTable}
              rowHeight={TABLE_ROW_HEIGHT}
              shouldHighlightRow={shouldHighlightTableRow}
              sortKey={sortKey}
              sortOrder={sortOrder}
              variants={renderedVariants}
            />
          ) : (
            <StatusMessage>No matching variants</StatusMessage>
          )}
        </Wrapper>
      </TrackPageSection>

      {showTableConfigurationModal && (
        <VariantTableConfigurationModal
          availableColumns={structuralVariantTableColumns}
          context={context}
          defaultColumns={DEFAULT_COLUMNS}
          selectedColumns={selectedColumns}
          onCancel={() => {
            setShowTableConfigurationModal(false)
          }}
          onSave={(newSelectedColumns) => {
            setSelectedColumns(newSelectedColumns)
            setShowTableConfigurationModal(false)

            userPreferences
              .savePreference('structuralVariantTableColumns', newSelectedColumns)
              .then(null, (error: any) => {
                showNotification({
                  title: 'Error',
                  message: error.message,
                  status: 'error',
                })
              })
          }}
        />
      )}
    </div>
  )
}

export default StructuralVariants
