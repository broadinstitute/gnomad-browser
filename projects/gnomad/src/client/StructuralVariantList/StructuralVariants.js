import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styled from 'styled-components'

import { PositionAxisTrack } from '@broad/region-viewer'
import { SegmentedControl } from '@broad/ui'
import { HUMAN_CHROMOSOMES } from '@broad/utilities'

import { TrackPageSection } from '../TrackPage'

import ExportStructuralVariantsButton from './ExportStructuralVariantsButton'
import filterVariants from './filterVariants'
import sortVariants from './sortVariants'
import {
  svConsequenceCategories,
  svConsequenceCategoryColors,
} from './structuralVariantConsequences'
import { svTypeColors } from './structuralVariantTypes'
import StructuralVariantFilterControls from './StructuralVariantFilterControls'
import StructrualVariantPropType from './StructuralVariantPropType'
import StructuralVariantsTable from './StructuralVariantsTable'
import StructuralVariantTracks from './StructuralVariantTracks'

const NUM_ROWS_RENDERED = 20
const TRACK_HEIGHT = 14
const TABLE_ROW_HEIGHT = 25

const Wrapper = styled.div`
  margin-bottom: 1em;
`

const ControlWrapper = styled(Wrapper)`
  display: flex;
  justify-content: flex-end;
  align-items: center;

  @media (min-width: 900px) {
    margin-right: 160px;
  }
`

class StructuralVariants extends Component {
  static propTypes = {
    chrom: PropTypes.string.isRequired,
    exportFileName: PropTypes.string.isRequired,
    variants: PropTypes.arrayOf(StructrualVariantPropType).isRequired,
    width: PropTypes.number.isRequired,
  }

  tracks = React.createRef()

  table = React.createRef()

  constructor(props) {
    super(props)

    const defaultFilter = {
      includeConsequenceCategories: {
        lof: true,
        dup_lof: true,
        copy_gain: true,
        other: true,
      },
      includeTypes: {
        DEL: true,
        DUP: true,
        MCNV: true,
        INS: true,
        INV: true,
        CPX: true,
        OTH: true,
      },
      includeFilteredVariants: false,
      searchText: '',
    }

    const defaultSortKey = 'pos'
    const defaultSortOrder = 'ascending'

    const renderedVariants = sortVariants(filterVariants(props.variants, defaultFilter), {
      sortKey: defaultSortKey,
      sortOrder: defaultSortOrder,
    })

    this.state = {
      filter: defaultFilter,
      highlightedVariantTrack: null,
      renderedVariants,
      shouldHighlightTableRow: () => false,
      sortKey: defaultSortKey,
      sortOrder: defaultSortOrder,
      colorKey: 'consequence',
    }
  }

  onFilter = newFilter => {
    this.setState(state => {
      const { variants } = this.props
      const { sortKey, sortOrder } = state
      const renderedVariants = sortVariants(filterVariants(variants, newFilter), {
        sortKey,
        sortOrder,
      })
      return {
        filter: newFilter,
        renderedVariants,
      }
    })
  }

  onSort = newSortKey => {
    this.setState(state => {
      const { renderedVariants, sortKey } = state

      let newSortOrder = 'descending'
      if (newSortKey === sortKey) {
        newSortOrder = state.sortOrder === 'ascending' ? 'descending' : 'ascending'
      }

      // Since the filter hasn't changed, sort the currently rendered variants instead
      // of filtering the input variants.
      const sortedVariants = sortVariants(renderedVariants, {
        sortKey: newSortKey,
        sortOrder: newSortOrder,
      })

      return {
        renderedVariants: sortedVariants,
        sortKey: newSortKey,
        sortOrder: newSortOrder,
      }
    })
  }

  onHoverVariantInTable = variantId => {
    this.setState({ highlightedVariantTrack: variantId })
  }

  onHoverVariantInTracks = variantId => {
    this.setState({
      highlightedVariantTrack: variantId,
      shouldHighlightTableRow: variantId
        ? variant => variant.variant_id === variantId
        : () => false,
    })
  }

  onScrollTable = ({ scrollOffset, scrollUpdateWasRequested }) => {
    if (this.tracks.current && !scrollUpdateWasRequested) {
      this.tracks.current.scrollTo(Math.round(scrollOffset * (TRACK_HEIGHT / TABLE_ROW_HEIGHT)))
    }
  }

  onScrollTracks = ({ scrollOffset, scrollUpdateWasRequested }) => {
    if (this.table.current && !scrollUpdateWasRequested) {
      this.table.current.scrollTo(Math.round(scrollOffset * (TABLE_ROW_HEIGHT / TRACK_HEIGHT)))
    }
  }

  trackColor = variant => {
    const { colorKey } = this.state
    if (colorKey === 'type') {
      return svTypeColors[variant.type] || svTypeColors.OTH
    }
    return variant.consequence
      ? svConsequenceCategoryColors[svConsequenceCategories[variant.consequence]]
      : svConsequenceCategoryColors.other
  }

  render() {
    const { chrom, exportFileName, width } = this.props
    const {
      filter,
      highlightedVariantTrack,
      renderedVariants,
      shouldHighlightTableRow,
      sortKey,
      sortOrder,
      colorKey,
    } = this.state

    const numRowsRendered = Math.min(renderedVariants.length, NUM_ROWS_RENDERED)

    // pos/end and pos2/end2 coordinates are based on the chromosome which they are located on.
    // If that chromosome is not the same as the one that the region viewer's coordinates
    // are based on, then offset the positions so that they are based on the
    // region viewer's coordinate system.
    const currentChromIndex = HUMAN_CHROMOSOMES.indexOf(chrom)
    const positionCorrectedVariants = renderedVariants.map(variant => {
      const copy = { ...variant }

      // This can only happen when chrom2/pos2/end2 is non-null
      if (variant.chrom2) {
        const chromIndex = HUMAN_CHROMOSOMES.indexOf(variant.chrom)
        const endChromIndex = HUMAN_CHROMOSOMES.indexOf(variant.chrom2)

        copy.pos += (chromIndex - currentChromIndex) * 1e9
        copy.end += (chromIndex - currentChromIndex) * 1e9

        copy.pos2 += (endChromIndex - currentChromIndex) * 1e9
        copy.end2 += (endChromIndex - currentChromIndex) * 1e9
      }

      return copy
    })

    return (
      <div>
        <ControlWrapper>
          <span style={{ marginRight: '0.5em' }}>Color variants by</span>
          <SegmentedControl
            id="sv-color-key"
            options={[
              { label: 'Consequence', value: 'consequence' },
              { label: 'Class', value: 'type' },
            ]}
            value={colorKey}
            onChange={value => {
              this.setState({ colorKey: value })
            }}
          />
        </ControlWrapper>
        <Wrapper>
          <StructuralVariantTracks
            ref={this.tracks}
            highlightedVariant={highlightedVariantTrack}
            numTracksRendered={numRowsRendered}
            onHover={this.onHoverVariantInTracks}
            onScroll={this.onScrollTracks}
            trackColor={this.trackColor}
            trackHeight={TRACK_HEIGHT}
            variants={positionCorrectedVariants}
          />
        </Wrapper>
        <Wrapper>
          <PositionAxisTrack />
        </Wrapper>
        <TrackPageSection style={{ fontSize: '14px' }}>
          <Wrapper>
            <StructuralVariantFilterControls
              colorKey={colorKey}
              value={filter}
              onChange={this.onFilter}
            />
            <div>
              <ExportStructuralVariantsButton
                exportFileName={exportFileName}
                variants={renderedVariants}
              />
            </div>
          </Wrapper>
          <Wrapper>
            <StructuralVariantsTable
              ref={this.table}
              cellData={{ colorKey, highlightWords: filter.searchText.split(/\s+/) }}
              numRowsRendered={numRowsRendered}
              onHoverVariant={this.onHoverVariantInTable}
              onRequestSort={this.onSort}
              onScroll={this.onScrollTable}
              rowHeight={TABLE_ROW_HEIGHT}
              shouldHighlightRow={shouldHighlightTableRow}
              sortKey={sortKey}
              sortOrder={sortOrder}
              variants={renderedVariants}
              width={width}
            />
          </Wrapper>
        </TrackPageSection>
      </div>
    )
  }
}

export default StructuralVariants
