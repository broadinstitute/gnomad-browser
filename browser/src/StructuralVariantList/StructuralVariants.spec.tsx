import React from 'react'
import { describe, test, expect } from '@jest/globals'
import StructuralVariants, { Context } from './StructuralVariants'
import structuralVariantFactory from '../__factories__/StructuralVariant'
import { createRenderer } from 'react-test-renderer/shallow'

describe('StructuralVariants', () => {
  const context: Context = { chrom: '12' }

  const variants = structuralVariantFactory.buildList(3)

  test('has no unexpected changes', () => {
    // We can't do a full render because the Grid component invoked a couple of
    // layers down uses react-sizeme. Since Grid is defined in GBTK, not this
    // project, the sizeme mock in this project isn't applied.
    const tree = createRenderer().render(
      <StructuralVariants context={context} exportFileName="somefile.tar.gz" variants={variants} />
    )
    expect(tree).toMatchSnapshot()
  })
})
