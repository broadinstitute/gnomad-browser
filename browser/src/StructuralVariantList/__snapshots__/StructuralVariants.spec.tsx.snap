// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructuralVariants has no unexpected changes 1`] = `
<div>
  <StructuralVariants__ControlWrapper>
    <span
      style={
        {
          "marginRight": "0.5em",
        }
      }
    >
      Color variants by
    </span>
    <SegmentedControl
      backgroundColor="#f8f9fa"
      borderColor="#6c757d"
      disabled={false}
      id="sv-color-key"
      onChange={[Function]}
      options={
        [
          {
            "label": "Consequence",
            "value": "consequence",
          },
          {
            "label": "Class",
            "value": "type",
          },
        ]
      }
      textColor="#000"
      value="type"
    />
  </StructuralVariants__ControlWrapper>
  <StructuralVariants__Wrapper>
    <ForwardRef
      highlightedVariant={null}
      numTracksRendered={3}
      onHover={[Function]}
      onScroll={[Function]}
      trackColor={[Function]}
      trackHeight={14}
      variants={
        [
          {
            "ac": 123,
            "ac_hemi": null,
            "ac_hom": null,
            "af": 456,
            "age_distribution": null,
            "algorithms": [],
            "alts": null,
            "an": 345,
            "chrom": "21",
            "chrom2": null,
            "consequence": null,
            "consequences": [],
            "copy_numbers": [],
            "cpx_intervals": [],
            "cpx_type": null,
            "end": 456,
            "end2": null,
            "evidence": [],
            "filters": [],
            "genes": [],
            "genotype_quality": null,
            "hemizygote_count": null,
            "homozygote_count": null,
            "length": 333,
            "major_consequence": null,
            "populations": [],
            "pos": 123,
            "pos2": null,
            "qual": 99,
            "type": "DUP",
            "variant_id": "DUP_21_999",
          },
          {
            "ac": 123,
            "ac_hemi": null,
            "ac_hom": null,
            "af": 456,
            "age_distribution": null,
            "algorithms": [],
            "alts": null,
            "an": 345,
            "chrom": "21",
            "chrom2": null,
            "consequence": null,
            "consequences": [],
            "copy_numbers": [],
            "cpx_intervals": [],
            "cpx_type": null,
            "end": 456,
            "end2": null,
            "evidence": [],
            "filters": [],
            "genes": [],
            "genotype_quality": null,
            "hemizygote_count": null,
            "homozygote_count": null,
            "length": 333,
            "major_consequence": null,
            "populations": [],
            "pos": 123,
            "pos2": null,
            "qual": 99,
            "type": "DUP",
            "variant_id": "DUP_21_999",
          },
          {
            "ac": 123,
            "ac_hemi": null,
            "ac_hom": null,
            "af": 456,
            "age_distribution": null,
            "algorithms": [],
            "alts": null,
            "an": 345,
            "chrom": "21",
            "chrom2": null,
            "consequence": null,
            "consequences": [],
            "copy_numbers": [],
            "cpx_intervals": [],
            "cpx_type": null,
            "end": 456,
            "end2": null,
            "evidence": [],
            "filters": [],
            "genes": [],
            "genotype_quality": null,
            "hemizygote_count": null,
            "homozygote_count": null,
            "length": 333,
            "major_consequence": null,
            "populations": [],
            "pos": 123,
            "pos2": null,
            "qual": 99,
            "type": "DUP",
            "variant_id": "DUP_21_999",
          },
        ]
      }
    />
  </StructuralVariants__Wrapper>
  <StructuralVariants__Wrapper>
    <PositionAxisTrack />
  </StructuralVariants__Wrapper>
  <TrackPage__TrackPageSection
    style={
      {
        "fontSize": "14px",
      }
    }
  >
    <StructuralVariants__Wrapper>
      <StructuralVariantFilterControls
        colorKey="type"
        onChange={[Function]}
        value={
          {
            "includeConsequenceCategories": {
              "copy_gain": true,
              "dup_lof": true,
              "lof": true,
              "other": true,
            },
            "includeFilteredVariants": false,
            "includeTypes": {
              "CPX": true,
              "DEL": true,
              "DUP": true,
              "INS": true,
              "INV": true,
              "MCNV": true,
              "OTH": true,
            },
            "searchText": "",
          }
        }
      />
      <div>
        <ExportStructuralVariantsButton
          exportFileName="somefile.tar.gz"
          variants={
            [
              {
                "ac": 123,
                "ac_hemi": null,
                "ac_hom": null,
                "af": 456,
                "age_distribution": null,
                "algorithms": [],
                "alts": null,
                "an": 345,
                "chrom": "21",
                "chrom2": null,
                "consequence": null,
                "consequences": [],
                "copy_numbers": [],
                "cpx_intervals": [],
                "cpx_type": null,
                "end": 456,
                "end2": null,
                "evidence": [],
                "filters": [],
                "genes": [],
                "genotype_quality": null,
                "hemizygote_count": null,
                "homozygote_count": null,
                "length": 333,
                "major_consequence": null,
                "populations": [],
                "pos": 123,
                "pos2": null,
                "qual": 99,
                "type": "DUP",
                "variant_id": "DUP_21_999",
              },
              {
                "ac": 123,
                "ac_hemi": null,
                "ac_hom": null,
                "af": 456,
                "age_distribution": null,
                "algorithms": [],
                "alts": null,
                "an": 345,
                "chrom": "21",
                "chrom2": null,
                "consequence": null,
                "consequences": [],
                "copy_numbers": [],
                "cpx_intervals": [],
                "cpx_type": null,
                "end": 456,
                "end2": null,
                "evidence": [],
                "filters": [],
                "genes": [],
                "genotype_quality": null,
                "hemizygote_count": null,
                "homozygote_count": null,
                "length": 333,
                "major_consequence": null,
                "populations": [],
                "pos": 123,
                "pos2": null,
                "qual": 99,
                "type": "DUP",
                "variant_id": "DUP_21_999",
              },
              {
                "ac": 123,
                "ac_hemi": null,
                "ac_hom": null,
                "af": 456,
                "age_distribution": null,
                "algorithms": [],
                "alts": null,
                "an": 345,
                "chrom": "21",
                "chrom2": null,
                "consequence": null,
                "consequences": [],
                "copy_numbers": [],
                "cpx_intervals": [],
                "cpx_type": null,
                "end": 456,
                "end2": null,
                "evidence": [],
                "filters": [],
                "genes": [],
                "genotype_quality": null,
                "hemizygote_count": null,
                "homozygote_count": null,
                "length": 333,
                "major_consequence": null,
                "populations": [],
                "pos": 123,
                "pos2": null,
                "qual": 99,
                "type": "DUP",
                "variant_id": "DUP_21_999",
              },
            ]
          }
        />
        <Button
          onClick={[Function]}
          style={
            {
              "marginLeft": "1ch",
            }
          }
        >
          Configure table
        </Button>
      </div>
    </StructuralVariants__Wrapper>
    <StructuralVariants__Wrapper
      style={
        {
          "minHeight": 115,
        }
      }
    >
      <ForwardRef
        cellData={
          {
            "colorKey": "type",
            "highlightWords": [
              "",
            ],
          }
        }
        columns={
          [
            {
              "compareFunction": [Function],
              "getSearchTerms": [Function],
              "heading": "Variant ID",
              "isRowHeader": true,
              "isSortable": true,
              "key": "variant_id",
              "minWidth": 110,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "grow": 0,
              "heading": "Source",
              "isSortable": false,
              "key": "source",
              "minWidth": 70,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "getSearchTerms": [Function],
              "heading": "Consequence",
              "isSortable": true,
              "key": "consequence",
              "minWidth": 160,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "getSearchTerms": [Function],
              "heading": "Class",
              "isSortable": true,
              "key": "class",
              "minWidth": 130,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "heading": "Position",
              "isSortable": true,
              "key": "pos",
              "minWidth": 200,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "heading": "Size",
              "isSortable": true,
              "key": "length",
              "minWidth": 100,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "heading": "Allele Count",
              "isSortable": true,
              "key": "ac",
              "minWidth": 110,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "heading": "Allele Number",
              "isSortable": true,
              "key": "an",
              "minWidth": 110,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "heading": "Allele Frequency",
              "isSortable": true,
              "key": "af",
              "minWidth": 110,
              "render": [Function],
              "tooltip": undefined,
            },
            {
              "compareFunction": [Function],
              "contextNotes": "Not shown when viewing Y chromosome",
              "heading": "Number of Homozygotes",
              "isSortable": true,
              "key": "homozygote_count",
              "minWidth": 100,
              "render": [Function],
              "shouldShowInContext": [Function],
              "tooltip": undefined,
            },
          ]
        }
        numRowsRendered={3}
        onHoverVariant={[Function]}
        onRequestSort={[Function]}
        onScroll={[Function]}
        rowHeight={25}
        shouldHighlightRow={[Function]}
        sortKey="variant_id"
        sortOrder="ascending"
        variants={
          [
            {
              "ac": 123,
              "ac_hemi": null,
              "ac_hom": null,
              "af": 456,
              "age_distribution": null,
              "algorithms": [],
              "alts": null,
              "an": 345,
              "chrom": "21",
              "chrom2": null,
              "consequence": null,
              "consequences": [],
              "copy_numbers": [],
              "cpx_intervals": [],
              "cpx_type": null,
              "end": 456,
              "end2": null,
              "evidence": [],
              "filters": [],
              "genes": [],
              "genotype_quality": null,
              "hemizygote_count": null,
              "homozygote_count": null,
              "length": 333,
              "major_consequence": null,
              "populations": [],
              "pos": 123,
              "pos2": null,
              "qual": 99,
              "type": "DUP",
              "variant_id": "DUP_21_999",
            },
            {
              "ac": 123,
              "ac_hemi": null,
              "ac_hom": null,
              "af": 456,
              "age_distribution": null,
              "algorithms": [],
              "alts": null,
              "an": 345,
              "chrom": "21",
              "chrom2": null,
              "consequence": null,
              "consequences": [],
              "copy_numbers": [],
              "cpx_intervals": [],
              "cpx_type": null,
              "end": 456,
              "end2": null,
              "evidence": [],
              "filters": [],
              "genes": [],
              "genotype_quality": null,
              "hemizygote_count": null,
              "homozygote_count": null,
              "length": 333,
              "major_consequence": null,
              "populations": [],
              "pos": 123,
              "pos2": null,
              "qual": 99,
              "type": "DUP",
              "variant_id": "DUP_21_999",
            },
            {
              "ac": 123,
              "ac_hemi": null,
              "ac_hom": null,
              "af": 456,
              "age_distribution": null,
              "algorithms": [],
              "alts": null,
              "an": 345,
              "chrom": "21",
              "chrom2": null,
              "consequence": null,
              "consequences": [],
              "copy_numbers": [],
              "cpx_intervals": [],
              "cpx_type": null,
              "end": 456,
              "end2": null,
              "evidence": [],
              "filters": [],
              "genes": [],
              "genotype_quality": null,
              "hemizygote_count": null,
              "homozygote_count": null,
              "length": 333,
              "major_consequence": null,
              "populations": [],
              "pos": 123,
              "pos2": null,
              "qual": 99,
              "type": "DUP",
              "variant_id": "DUP_21_999",
            },
          ]
        }
      />
    </StructuralVariants__Wrapper>
  </TrackPage__TrackPageSection>
</div>
`;
